### ISSUE TITLE
Checking De-Bounce Code

### ISSUE THREAD

**Issue by JMenterZXT** [madvosburg/Pendant_Project_Master#12]
https://github.com/madvosburg/Pendant_Project_Master/blob/b4d85dc184df4f4d764c78ea90df50d3ad2de04f/Core/Src/main.c#L488

Trying out issue creation. Maybe review is the better way to do this.

I've got one question and a note which will lead to fully vetting this implementation:
- Is this EXT1 GPIO callback set to a single edge only? Rising or falling edge? So that it only triggers when button is pressed. I assume it's only on the "pressed" edge and not released.
- Once I know this I will think about if comparing to the previous state makes sense. If it's single edge I'd think you only to need to check that it is still in the correct state (pressed state), not that it matches the previous state.

**Comment by JMenterZXT** [comment]
Forgot the @madvosburg, to make sure the notification happens.

**Comment by JMenterZXT** [comment]
Ok, I thought about it a little bit more, and this interrupt for EXT1 GPIO must be set to trigger on both edges, and that's why it needs to track the actual initial state and compare it to the state after the timer interrupt fires.
This seems pretty solid, but I think the potential for calling the timer start function multiple times while the switch is bouncing could possibly be an issue. I'd propose creating a new 32-bit variable that holds flags for each button, the flag is set for the corresponding button on the EXT1 call-back and cleared when getting through the final debounce logic to set the button bitmask. The purpose of the flag is to block the timer 16 start function from being called for that same button while the timer is running. So  it'd be an if statement checking that flag (to call timer 16 start) inside the if statement for the corresponding button in the EXT1 callback. Either that or you could just check a variable that the timer is running (for example counter value is not 0).

I think that's worth a shot. If the makes any sense let me know, otherwise I could post a code snippet.

**Comment by madvosburg** [comment]
uint32_t button_flag = 0;

void HAL_GPIO_EXTI_Callback(uint16_t GPIO_Pin)
{
  if (GPIO_Pin == GPIO_PIN_2){
	  state = HAL_GPIO_ReadPin(GPIOC, GPIO_PIN_2);
	  button = BUTTON_1;	//red
	  if(!bitmask_check(button_flag, 0)){
		  HAL_TIM_Base_Start_IT(&htim16);
		  button_flag = bitmask_set(button_flag, 0);
	  }
  }
}
void HAL_TIM_PeriodElapsedCallback(TIM_HandleTypeDef *htim){
	if(htim == &htim16){
		if(HAL_GPIO_ReadPin(GPIOC, GPIO_PIN_2) == state && button == BUTTON_1){
			bit = bitmask_set(bit, 0);
			button_flag = bitmask_clear(button_flag, 0);
		}
        }
}

**Comment by madvosburg** [comment]
Here is what I've come up with. This is just a snippet for the first button but it's similar for the others. Let me know if that's what you were suggesting @JMenterZXT

**Comment by JMenterZXT** [comment]
@madvosburg Yep, that's exactly right for using another 32-bit flag variable to block the timer start. Let me know if that makes any difference on the random extra state changes.

**Comment by madvosburg** [comment]
Just tested with the new code, unfortunately it didn't seem to fix it. Any other ideas as to why it might be tracking extra presses?

**Comment by JMenterZXT** [comment]
I'd try out an RC filter to see if it is actually contact bounce causing the issue. If you can get some through-hole resistor and capacitors, that could be wired in pretty easily with a breadboard. The resistor goes in series from button to the GPIO pin, the capacitor from GPIO pin to ground. The time constant can be just about anything over ~50ms, even 500ms would probably be ok (T=RC). 
Do you have a capacitor in the uF range (2.2-47 uF) you can use for that?  A 10k with 47uF would be 470ms time constant.

**Comment by madvosburg** [comment]
I have a 0.1uF and 10uF capacitor. I can try wiring up the 10uF and see if that helps
